File: leetcode/110.BalancedBinaryTree.go
package leetcode

import "math"

type Traverse struct {
	Height     int
	IsBalanced bool
}

func isBalanced(root *TreeNode) bool {
	return traverse(root).IsBalanced
}
func traverse(node *TreeNode) Traverse {
	if node == nil {
		return Traverse{Height: 0, IsBalanced: true}
	}

	left := traverse(node.Left)
	right := traverse(node.Right)
	b := left.IsBalanced && right.IsBalanced && math.Abs(float64(left.Height-right.Height)) <= 1

	return Traverse{
		IsBalanced: b,
		Height:     1 + int(math.Max(float64(left.Height), float64(right.Height))),
	}
}

//tree := &TreeNode{Val: 3}
//tree.Left = &TreeNode{Val: 9}
//
//tree.Right = &TreeNode{Val: 20}
//tree.Right.Left = &TreeNode{Val: 15}
//tree.Right.Right = &TreeNode{Val: 7}
//
//result := isBalanced(tree)
//
//fmt.Printf("%+v", result)

File: leetcode/110.BalancedBinaryTree.ts
import { TreeNode } from './shared'

/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */
function isBalanced(root: TreeNode | null): boolean {
  const dfs = (
    node: TreeNode | null,
  ): { isBalanced: boolean; height: number } => {
    if (!node) {
      return {
        isBalanced: true,
        height: 0,
      }
    }

    const left = dfs(node.left)
    const right = dfs(node.right)
    const isBalanced =
      left?.isBalanced &&
      right?.isBalanced &&
      Math.abs(left.height - right.height) <= 1

    return {
      isBalanced,
      height: 1 + Math.max(left.height, right.height),
    }
  }

  return dfs(root).isBalanced
}

const tree = new TreeNode(3)
tree.left = new TreeNode(9)
tree.right = new TreeNode(20)
tree.right.left = new TreeNode(15)
tree.right.right = new TreeNode(7)

console.log(isBalanced(tree))
